implement dropArgumentsToMatch method

fengxue-IS opened the pull request:
added implementation for MethodHandle.dropArgumentsToMatch(...)
added testcode to check the output of method, modified testng suite
to configure and run the new test cases

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

fengxue-IS commented:
reviewer: @tajila 
test review: @TianyuZuo

genie-openj9 commented:
Can one of the admins verify this patch?

tajila commented on test/Jsr292/src/com/ibm/j9/jsr292/api/MethodHandleAPI_dropArgumentsToMatch.java:
```diff
@@ -0,0 +1,399 @@
+
+		Assert.fail("dropArgumentsToMatch method is nut supposed to accept null parameter list");
+	}
+
```
A test with a target methodHandle that accepts no parameters (ie ()Ljava/lang/String;) might be usefull as well. 

Similarly a non-NULL empty parameter list would also be usefull

tajila commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+			throw new IllegalArgumentException("Invalid parameters: skippedArgumentCount"); //$NON-NLS-1$
+		}
+		if ((0 > location) ||
+			(valueTypes.size() < location) ||
```
if location is zero indexed then wouldn't "valueTypes.size() <= location" also be out of range?

tajila replied:
I discussed with Jack and the second check ensures that location never overflows

TianyuZuo commented on test/Jsr292/src/com/ibm/j9/jsr292/api/MethodHandleAPI_dropArgumentsToMatch.java:
```diff
@@ -0,0 +1,399 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2017 IBM Corp. and others
```
If this is a new introduced file, the year should begin from 2017

pshipton commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+		/* check if indexing is in range */
+		if ((0 > skippedArgumentCount) ||
+			(skippedArgumentCount > originalType.parameterCount())) {
+			throw new IllegalArgumentException("Invalid parameters: skippedArgumentCount"); //$NON-NLS-1$
```
The string should be externalized. Look for MSG in this class to find examples.

fengxue-IS replied:
fixed, moved exceptions to ExternalMesages-MasterIndex
added entries K0670 - K0674

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+		Class<?>[] valueTypesCopy = new Class<?>[valueTypes.size()];
+
+		/* check if indexing is in range */
+		if ((0 > skippedArgumentCount) ||
```
Can you write this as
```
if ((skippedArgumentCount < 0) || 
```
It tends to be easier to read as the bounds then.

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+		/* check if indexing is in range */
+		if ((0 > skippedArgumentCount) ||
+			(skippedArgumentCount > originalType.parameterCount())) {
+			throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0670")); //$NON-NLS-1$
```
This should be proceeded by a comment that includes the K0670 and the message using preprocessor tags.  Search for other uses of the com.ibm.oti.util.Msg class to see how the comment should look

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+	public static final  MethodHandle dropArgumentsToMatch(MethodHandle originalHandle, int skippedArgumentCount, List<Class<?>> valueTypes, int location) {
+		/* implicit null checks */
+		MethodType originalType = originalHandle.type;
+		Class<?>[] valueTypesCopy = new Class<?>[valueTypes.size()];
```
Please cache `valueTypes.size()` in a local.  Because the parameter is a List, we don't know how expensive that operation might be and the result is used repeatedly in this method

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+			(skippedArgumentCount > originalType.parameterCount())) {
+			throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0670")); //$NON-NLS-1$
+		}
+		if ((0 > location) ||
```
`0 > value` is a strange idiom for checking if the value is negative.  Write the code so it's easy to read and matches most readers expectations.

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+		if ((0 > location) ||
+			(valueTypes.size() < location) ||
+			(valueTypes.size() < location + originalType.parameterCount() - skippedArgumentCount)) {
+			throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0671")); //$NON-NLS-1$
```
This should be proceeded by a comment that includes the K0670 and the message using preprocessor tags.  Search for other uses of the com.ibm.oti.util.Msg class to see how the comment should look

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+		/* check for void.class in list during clone process */
+		for (int i = 0; i < valueTypes.size(); i++) {
+			if (valueTypes.get(i) == void.class) {
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0672")); //$NON-NLS-1$
```
This should be proceeded by a comment that includes the K0670 and the message using preprocessor tags.  Search for other uses of the com.ibm.oti.util.Msg class to see how the comment should look

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0672")); //$NON-NLS-1$
+			}
+			else if (valueTypes.get(i) == null) {
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0673")); //$NON-NLS-1$
```
This should be proceeded by a comment that includes the K0670 and the message using preprocessor tags.  Search for other uses of the com.ibm.oti.util.Msg class to see how the comment should look

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+		/* check if sublist match */
+		for (int i = skippedArgumentCount; i < ptypes.length; i++) {
+			if (ptypes[i] != valueTypesCopy[i + location - skippedArgumentCount]) {
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0674")); //$NON-NLS-1$
```
This should be proceeded by a comment that includes the K0670 and the message using preprocessor tags.  Search for other uses of the com.ibm.oti.util.Msg class to see how the comment should look

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+		}
+
+		/* check for void.class in list during clone process */
+		for (int i = 0; i < valueTypes.size(); i++) {
```
cache the result of `valueTypes.get(i)` in a local in the for loop so it only needs to be executed once.

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2818,11 +2818,62 @@ public static MethodHandle dropArguments(MethodHandle originalHandle, int locati
+		for (int i = 0; i < valueTypes.size(); i++) {
+			if (valueTypes.get(i) == void.class) {
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0672")); //$NON-NLS-1$
+			}
```
Formatting nitpick: please write this as
``` 
} else if {
```
Here and other places need to be modified.

fengxue-IS commented:
fixed as Dan suggested,
I changed the List to Array copy to use `List.toArray(T[])` so index access uses `array[index]` which I assume would be constant time

DanHeidinga commented:
Jenkins test sanity

AdamBrousseau commented:
Jenkins test sanity
